### PR TITLE
In this example I read and write json file using Jackson library

### PR DESCRIPTION
**Pull request by manolovv**

### CODE REVIEW

File: src/main/java/com/manolov/Main.java [hunk 1]
```diff
@@ -1,11 +1,29 @@
...
 
 public class Main {
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
+
+
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.findAndRegisterModules();
+
+        List<Vehicle> carList = mapper.readValue(new File("src\\main\\resources\\vehicles.json"),
```
Comment (xerZV) [inline src/main/java/com/manolov/Main.java]: this is a collection full of vehicles, rename the field properly

File: src/main/java/com/manolov/Main.java [hunk 2]
```diff
@@ -31,6 +47,18 @@ public static void main(String[] args) {
         Motor hondaCBR = new Motor(hondaCBRManufacture, hondaCBRRegister, hondaCBRBrand, hondaCBRModel, hondaCBRFeatures,
                 hondaCBREngine, TransmissionChoice.MANUAL, EuroStandard.EURO_4, 24000, Color.YELLOW, MotorCategory.ENDURO);
 
-        System.out.println(hondaCBR);
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
Comment (xerZV) [inline src/main/java/com/manolov/Main.java]: you do not need second `ObjectMapper` instance, use the `ObjectMapper` instance created on line 17

File: src/main/java/com/manolov/architecture/Car.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package com.manolov.architecture;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.time.LocalDate;
+import java.util.List;
+
+public class Car extends Vehicle {
+    private final String type = "Car";
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Car.java]: this field can be constant

File: src/main/java/com/manolov/architecture/Motor.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package com.manolov.architecture;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.time.LocalDate;
+import java.util.List;
+
+public class Motor extends Vehicle {
+    private final String type = "Motor";
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Motor.java]: this field can be constant

File: src/main/resources/writeObjects.json
Comment (xerZV) [inline src/main/resources/writeObjects.json:1]: delete this file from the `resources` folder and add the file name in `.gitignore` file so next time you run the application the git will ignore this file and wont bring it to the remote.

**Review by xerZV — CHANGES_REQUESTED** [review]
remove whole .idea folder from the project and commit/push. It is added in the .gitignore file but maybe after pushing the folder. It is hard for other contributors to work with your IDE workspace settings.

Overall great work, minimal changes required.
Keep up the great work, @manolovv

File: src/main/java/com/manolov/Main.java [hunk 5]
```diff
@@ -1,11 +1,29 @@
...
 public class Main {
-    public static void main(String[] args) {
+    public static void main(String[] args) throws IOException {
+
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.findAndRegisterModules();
+
+        List<Vehicle> vehicleListReader = objectMapper.readValue(new File("src\\main\\resources\\vehicles.json"),
+                new TypeReference<List<Vehicle>>() {
```
Comment (xerZV) [inline src/main/java/com/manolov/Main.java]: remove redundant `List<Vehicle>` in the `TypeReference`

File: src/main/java/com/manolov/Main.java [hunk 6]
```diff
@@ -31,6 +47,27 @@ public static void main(String[] args) {
         Motor hondaCBR = new Motor(hondaCBRManufacture, hondaCBRRegister, hondaCBRBrand, hondaCBRModel, hondaCBRFeatures,
                 hondaCBREngine, TransmissionChoice.MANUAL, EuroStandard.EURO_4, 24000, Color.YELLOW, MotorCategory.ENDURO);
 
-        System.out.println(hondaCBR);
+
+        objectMapper.findAndRegisterModules();
+        SimpleDateFormat df = new SimpleDateFormat("dd-MM-yyyy");
```
Comment (xerZV) [inline src/main/java/com/manolov/Main.java]: duplicates `objectMapper.findAndRegisterModules();` - you have same thing in line 19

File: src/main/java/com/manolov/Main.java [hunk 7]
```diff
@@ -31,6 +47,27 @@ public static void main(String[] args) {
...
+        objectMapper.setDateFormat(df);
+
+        List<Vehicle> vehicleListWriter = new ArrayList<>();
+        vehicleListWriter.add(audiR8);
+        vehicleListWriter.add(hondaCBR);
+
+        String json = objectMapper.writeValueAsString(vehicleListWriter);
+
+
```
Comment (xerZV) [inline src/main/java/com/manolov/Main.java:60]: Is there a reason you deleted the line on which you save to JSON file?

File: src/main/java/com/manolov/architecture/Ad.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+import java.math.BigDecimal;
+
+public class Ad {
+    private final Car car;
+    private final Motor motor;
+    private final ProductStatus productStatus;
+    private final User user;
+    private final String phoneNumber;
+    private final BigDecimal price;
+    private TextField moreInformation;
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java]: use `String` instead

File: src/main/java/com/manolov/architecture/Ad.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.awt.*;
+import java.math.BigDecimal;
+
+public class Ad {
+    private final Car car;
+    private final Motor motor;
+    private final ProductStatus productStatus;
+    private final User user;
+    private final String phoneNumber;
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java]: you can move this field to `User` class

File: src/main/java/com/manolov/architecture/Ad.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    public Ad(Car car, ProductStatus productStatus,
+              User user, String phoneNumber,
+              BigDecimal price, TextField moreInformation) {
+        this.motor = null;
+        this.car = car;
+        this.productStatus = productStatus;
+        this.user = user;
+        this.phoneNumber = phoneNumber;
+        this.price = price;
+        if (moreInformation != null) {
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java]: unnecessary if statement, just to `this.moreInformation = moreInformation`

File: src/main/java/com/manolov/architecture/Ad.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    public Ad(Motor motor, ProductStatus productStatus,
+              User user, String phoneNumber,
+              BigDecimal price, TextField moreInformation) {
+        this.car = null;
+        this.motor = motor;
+        this.productStatus = productStatus;
+        this.user = user;
+        this.phoneNumber = phoneNumber;
+        this.price = price;
+        if (moreInformation != null) {
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java]: unnecessary if statement, just to `this.moreInformation = moreInformation`

**Review by xerZV — CHANGES_REQUESTED** [review]

File: src/main/java/com/manolov/Main.java (hunk 1 above)
Comment (manolovv) [reply src/main/java/com/manolov/Main.java]: I`ve done that

File: src/main/java/com/manolov/Main.java (hunk 5 above)
Comment (manolovv) [reply src/main/java/com/manolov/Main.java]: its done

File: src/main/java/com/manolov/architecture/Ad.java (hunk 9 above)
Comment (manolovv) [reply src/main/java/com/manolov/architecture/Ad.java]: I must to remove the Ad class ?

File: src/main/java/com/manolov/Main.java (hunk 7 above)
Comment (manolovv) [reply src/main/java/com/manolov/Main.java:60]: You said to remove the json file in which i recorded info about the objects. Now i write in this json String

File: src/main/java/com/manolov/architecture/Ad.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
+package com.manolov.architecture;
+
+import java.awt.*;
+import java.math.BigDecimal;
+
+public class Ad {
+    private final Car car;
+    private final Motor motor;
+    private final ProductStatus productStatus;
+    private final User user;
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java:9]: just a test comment on multiple lines so you can see how it looks

File: src/main/java/com/manolov/architecture/Ad.java (hunk 9 above)
Comment (xerZV) [reply src/main/java/com/manolov/architecture/Ad.java]: you can move the `phoneNumber` field from `Ad` class to `User` class

File: src/main/java/com/manolov/architecture/Ad.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
+package com.manolov.architecture;
+
+import java.awt.*;
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Ad.java]: Unused import statement

File: src/main/java/com/manolov/architecture/Car.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
+package com.manolov.architecture;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.time.LocalDate;
+import java.util.List;
+
+public class Car extends Vehicle {
+    private static final String CAR_TYPE = "Car";
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Car.java]: this private field is never used - you can remove it

File: src/main/java/com/manolov/architecture/Motor.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
+package com.manolov.architecture;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.time.LocalDate;
+import java.util.List;
+
+public class Motor extends Vehicle {
+    private static final String MOTOR_TYPE = "Motor";
```
Comment (xerZV) [inline src/main/java/com/manolov/architecture/Motor.java]: this private field is never used - you can remove it

File: src/main/java/com/manolov/architecture/Ad.java (hunk 11 above)
Comment (manolovv) [reply src/main/java/com/manolov/architecture/Ad.java]: its already done

**Review by xerZV — APPROVED** [review]

**Pull request merged**